azurelinuxagent/common/protocol/goal_state.py

```diff
@@ -128,40 +122,64 @@ def update_host_plugin_headers(wire_client):
-        else:
-            # else ensure the extensions are using the latest vm_settings
-            timestamp = create_timestamp()
-            vm_settings, vm_settings_updated = self._fetch_vm_settings(force_update=force_update)
+        #
+        # Fetch the goal state from both the HGAP and the WireServer
+        #
+        timestamp = timeutil.create_timestamp()
+
+        incarnation, xml_text, xml_doc = GoalState._fetch_goal_state(self._wire_client)
+        goal_state_updated = incarnation != self._incarnation
+        if goal_state_updated:
+            logger.info('Fetched new goal state from the WireServer [incarnation {0}]', incarnation)
+
+        vm_settings, vm_settings_updated = GoalState._fetch_vm_settings(self._wire_client)
+        if vm_settings_updated:
+            logger.info("Fetched new vmSettings [HostGAPlugin correlation ID: {0} eTag: {1} source: {2}]", vm_settings.hostga_plugin_correlation_id, vm_settings.etag, vm_settings.source)
+        # Ignore the vmSettings if their source is Fabric (processing a Fabric goal state may require the tenant certificate and the vmSettings don't include it.)
+        if vm_settings is not None and vm_settings.source == GoalStateSource.Fabric:
             if vm_settings_updated:
-                logger.info("Fetched new vmSettings [HostGAPlugin correlation ID: {0} eTag: {1} source: {2}]", vm_settings.hostga_plugin_correlation_id, vm_settings.etag, vm_settings.source)
-                self._history = GoalStateHistory(timestamp, vm_settings.etag)
-                self._extensions_goal_state = vm_settings
-                self._history.save_vm_settings(vm_settings.get_redacted_text())
+                logger.info("The vmSettings originated via Fabric; will ignore them.")
+            vm_settings, vm_settings_updated = None, False
+
+        # If neither goal state has changed we are done with the update
+        if not goal_state_updated and not vm_settings_updated:
+            return
+
+        # Start a new history subdirectory and capture the updated goal state
+        tag = "{0}".format(incarnation) if vm_settings is None else "{0}-{1}".format(incarnation, vm_settings.etag)
+        self._history = GoalStateHistory(timestamp, tag)
+        if goal_state_updated:
+            self._history.save_goal_state(xml_text)
+        if vm_settings_updated:
+            self._history.save_vm_settings(vm_settings.get_redacted_text())
+
+        #
+        # Continue fetching the rest of the goal state
+        #
+        extensions_config = None
+        if goal_state_updated:
+            extensions_config = self._fetch_full_wire_server_goal_state(incarnation, xml_doc)
+
+        #
+        # Lastly, decide whether to use the vmSettings or extensionsConfig for the extensions goal state
+        #
+        if goal_state_updated and vm_settings_updated:
+            most_recent = vm_settings if vm_settings.created_on_timestamp > extensions_config.created_on_timestamp else extensions_config
```

Comment:
Don't we process whichever timestamp is older? Otherwise, order becomes latest first and older next right?

Consider a case we have 1 hour gap between our goal state processing, in that period we have wireserver goal state updated and then FT `vm_seetings` updated. Now we have these two when we process current goal state and according to `created_timestamp` wireserver goal state is older. So, don't we process wire server first? take `extension_config` instead of `vm_seetings`?

Reply:
We always want to execute the latest goal state, and skipping older goal states is fine. The latest goal state will contain the desired state of the VM and the agent only needs to apply that one.